README
PerlMol version 0.3200
======================

PerlMol is a collection of Perl modules for chemoinformatics and computational
chemistry with the philosophy that "simple things should be simple". It should
be possible to write one-liners that use this toolkit to do meaningful
"molecular munging". The perlmol toolkit provides objects and methods for
representing molecules, atoms, and bonds in Perl; doing substructure matching;
and reading and writing files in various formats.

This is a bundle containing all of the modules of the PerlMol Project and their
dependencies.

CONTENTS

    Chemistry-Bond-Find             0.21
    Chemistry-Canonicalize          0.10
    Chemistry-File-MDLMol           0.19
    Chemistry-File-Mopac            0.15
    Chemistry-File-PDB              0.20
    Chemistry-File-SLN              0.10
    Chemistry-File-SMARTS           0.21
    Chemistry-File-SMILES           0.43
    Chemistry-File-XYZ              0.11
    Chemistry-FormulaPattern        0.10
    Chemistry-InternalCoords        0.18
    Chemistry-Isotope               0.10
    Chemistry-MacroMol              0.06
    Chemistry-MidasPattern          0.10
    Chemistry-Mok                   0.24
    Chemistry-Mol                   0.32
    Chemistry-Pattern               0.25
    Chemistry-Reaction              0.02
    Chemistry-Ring                  0.18
    Math-VectorReal                 1.02
    Parse-Yapp                      1.05
    Statistics-Regression           0.15


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

    perl-5.6.0 or a more recent version.

If you have perl-5.8.0 (or a more recent version), this bundle has everything
you need. If you have an earlier version of Perl, you also need these core
modules:

    Test::Simple        0.45
    Scalar::Util        1.07
    Storable            2.04
    Text::Balanced      1.89

COPYRIGHT AND LICENSE

Copyright (C) 2005 Ivan Tubert-Brohman <itub@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for PerlMol

0.3500  May 24 2005
    NEW!!! Added an examples/ directory

    New distributions:
        Chemistry-File-VRML             0.10
        Chemistry-3DBuilder             0.10

    Changed versions:

        Chemistry-Mol                   0.35
            - Added formal_radical atom property.
            - Compatibility with Storable-2.14, which already takes care of
              weak references.
            - calc_implicit_hydrogens, add_implicit_hydrogens
            - Fixed $atom->bonds duplication on $mol->separate (bug 1173237)
            - New method: safe_clone
            - Fixed spurious warnings in sprout_hydrogens (bug 1157393)
            - Added a test for incompatible Chemistry::File::SMILES versions.

        Chemistry-Mok                   0.25
            - Read one molecule at a time using the Chemistry-Mol-0.30 
              interface.
            - New special variable $FH
            - Added support for Chemistry::3DBuilder (-3 command-line option)
            - Add #line to the generated code so that errors give more 
              meaningful line numbers.
            - Fixed parsing bug where the first block used an explicit pattern
              language
            - Restructured Chemistry::Mok to be more object-oriented: 
            - WARNING: the mok object is now a hash reference instead of an 
              array reference. This may affect derived classes.

        Chemistry-File-SMILES           0.44
            - Fixed memory leak that prevented the last atom from being garbage
              collected.

        Chemistry-File-SMARTS           0.22
            - Added the original SMARTS string as the $patt->name property.
            - Added support for atom mapping numbers

        Chemistry-File-PDB              0.21
            - Changed parse_file to read_mol
            - Read multi-model files (each model read as one molecule)
            - Read chain id and insertion code

        Chemistry-Pattern               0.26
            - Fixed mapping bug for overlap => 0

        Chemistry-MidasPattern          0.11
            - Chain support.
            - Be more lenient with whitespace.
            - Refactored parsing and execution.

        Chemistry-File-MDLMol           0.20
            - Added support for charges and radicals.
            - Add implicit hydrogens on read.
            - Support for some query properties: atom lists, bond topology, and
              bond types.

        Chemistry-Isotope               0.11
            - Added natural abundance table

    Unchanged versions:
        Chemistry-InternalCoords        0.18
        Chemistry-MacroMol              0.06
        Chemistry-Ring                  0.18
        Chemistry-Canonicalize          0.10
        Chemistry-Bond-Find             0.21
        Chemistry-File-SLN              0.10
        Chemistry-FormulaPattern        0.10
        Chemistry-File-Mopac            0.15
        Chemistry-File-XYZ              0.11
        Chemistry-Reaction              0.02
        Math-VectorReal                 1.02
        Statistics-Regression           0.15
        Parse-Yapp                      1.05

0.3200  Mar 04 2005
    New distributions:
        Chemistry-Reaction              0.02
        Chemistry-Isotope               0.10
    Changed versions:
        Chemistry-Mol                   0.32
        Chemistry-Mok                   0.24
        Chemistry-InternalCoords        0.18
        Chemistry-File-MDLMol           0.19
        Chemistry-File-SMILES           0.43
        Chemistry-File-SMARTS           0.21
    Unchanged versions:
        Chemistry-MacroMol              0.06
        Chemistry-Ring                  0.18
        Chemistry-Canonicalize          0.10
        Chemistry-Pattern               0.25
        Chemistry-Bond-Find             0.21
        Chemistry-File-SLN              0.10
        Chemistry-File-PDB              0.20
        Chemistry-FormulaPattern        0.10
        Chemistry-MidasPattern          0.10
        Chemistry-File-Mopac            0.15
        Chemistry-File-XYZ              0.11
        Math-VectorReal                 1.02
        Statistics-Regression           0.15
        Parse-Yapp                      1.05

0.2600  Aug 12 2004
    New distributions:
        Chemistry-File-SLN              0.10
        Chemistry-FormulaPattern        0.10
        Parse-Yapp                      1.05
    Changed versions:
        Chemistry-Mol                   0.26
        Chemistry-InternalCoords        0.17
        Chemistry-Ring                  0.18
        Chemistry-Pattern               0.25
        Chemistry-File-MDLMol           0.17
        Chemistry-File-SMILES           0.42
        Chemistry-File-SMARTS           0.20
        Chemistry-File-XYZ              0.11
        Chemistry-Mok                   0.23
    Unchanged versions:
        Chemistry-MacroMol              0.06
        Chemistry-Canonicalize          0.10
        Chemistry-Bond-Find             0.21
        Chemistry-File-PDB              0.20
        Chemistry-MidasPattern          0.10
        Chemistry-File-Mopac            0.15
        Math-VectorReal                 1.02
        Statistics-Regression           0.15


0.2500  Jul 03 2004
    New distributions:
        Chemistry-File-XYZ              0.10
        Chemistry-MidasPattern          0.10
    Changed versions:
        Chemistry-File-MDLMol           0.16
        Chemistry-File-Mopac            0.15
        Chemistry-File-PDB              0.20
        Chemistry-File-SMARTS           0.11
        Chemistry-File-SMILES           0.41
        Chemistry-InternalCoords        0.16
        Chemistry-MacroMol              0.06
        Chemistry-Mok                   0.22
        Chemistry-Mol                   0.25
        Chemistry-Ring                  0.15
    Unchanged versions:
        Chemistry-Bond-Find             0.21
        Chemistry-Canonicalize          0.10

examples/file_conversion/pbabel.pl
#!/home/ivan/bin/perl

# This sample program is called 'pbabel' because it is inspired by the Babel
# program and it is written in Perl using PerlMol.

# USAGE
# ./pbabel.pl [options] <input file> <output file>

use strict;
use warnings;

# load all File I/O modules
use Chemistry::File ':auto';

# load helper modules for finding bonds, aromatizing, and generating coords
use Chemistry::3DBuilder 'build_3d';
use Chemistry::Ring 'aromatize_mol';
use Chemistry::Bond::Find 'find_bonds';

# parse options
use Getopt::Long;
Getopt::Long::Configure ("bundling");
my ($input_format, $output_format, $build_3d, $find_bonds, $aromatize,
    $no_kekulize, $unique, $name); 
my $result = GetOptions(
    "i=s"    => \$input_format,
    "o=s"    => \$output_format,
    "3"      => \$build_3d,
    "b"      => \$find_bonds,
    "a"      => \$aromatize,
    "K"      => \$no_kekulize,
    "u"      => \$unique,
    "n"      => \$name,
);

# check if everything is OK, or else print usage info
if (@ARGV != 2 or !$result) {
    my @formats = Chemistry::Mol->formats;
    print <<OPTIONS;
$0 [options] <input file> <output file>

Options:
    -i <input format>
    -o <output format>
    -3      build 3d coordinates 
    -b      find bonds
    -a      aromatize
    -K      don't kekulize (only for reading SMILES)
    -u      unique (only for writing SMILES)
    -n      include name (only for writing SMILES)

Available file formats (if omitted they will be guessed):
OPTIONS
print "    $_\n" for @formats;
    exit;
}

my ($input_file, $output_file) = @ARGV;

# read the input file
my @mols = Chemistry::Mol->read($input_file, 
    format   => $input_format,
    kekulize => ! $no_kekulize,  # only used by SMILES
);

# do optional procession on the molecules
for my $mol (@mols) {
    build_3d($mol)                  if $build_3d;
    find_bonds($mol, orders => 1)   if $find_bonds;
    aromatize_mol($mol)             if $aromatize;
}

# write the output file
$mols[0]->write($output_file, 
    format => $output_format, 
    mols => \@mols, # only used by multi-molecule files such as SMILES and SDF
    name => $name,              # only used by SMILES
    unique => $unique,          # only used by SMILES
);



examples/peptide_builder/peptide_builder.pl
#!/home/ivan/bin/perl
# usage: ./pep.pl <SEQUENCE>
# prints an mdl molfile to stdout

# curent version only supports the residues A,G,L,V
# the residues are assumed to be in the files listed below in %files,
# which must have the atoms N, CA, and C as the first three atoms in the 
# file

use strict;
use warnings;
#use diagnostics;

use Chemistry::File::MDLMol;
use Chemistry::File::SMARTS;
use Chemistry::Pattern;
use Chemistry::InternalCoords::Builder qw(build_zmat);

{
    my %files = (
        A => 'ala.mol',
        V => 'val.mol',
        G => 'gly.mol',
        L => 'leu.mol',
    );

    my $patt = Chemistry::Pattern->parse('C((=O)CN([C,H])[H])O[H]', 
        format => 'smarts');

    my %aa;

    my ($c, $o, $ca, $n, $hn);

    sub amino_acid {
        my ($code) = @_;
        return $aa{$code} if $aa{$code};
        my $file = $files{$code} or die "unknown code '$code', stopping ";
        $aa{$code} = Chemistry::Mol->read($file);
    }

    sub start_chain {
        my ($res) = @_;
        my $chain = safe_clone($res);
        $patt->match($chain) or die "didn't match, stopping ";
        my ($oh, $ho, $hnb);
        ($c, $o, $ca, $n, $hn, $hnb, $oh, $ho) = $patt->atom_map;
        $oh->delete; $ho->delete;
        build_zmat($chain, bfs => 0);
        $chain;
    }

    sub add_residue {
        my ($chain, $res) = @_;
        $res = safe_clone($res);

        $patt->match($res) or die "didn't match, stopping ";
        my @map = $patt->atom_map;
        $_->delete for splice @map, 5;
        build_zmat($res, bfs => 0);

        $chain->combine($res);
        my ($c2, $o2, $ca2, $n2, $hn2) = 
            map { $chain->by_id($_->id) } @map;

        $chain->new_bond(atoms => [$n2, $c], order => 1);
        $n2 ->internal_coords($c, 1.5, $ca, 120, $o, 180);
        $ca2->internal_coords($n2, 1.5, $c, 120, $ca, 180);
        $c2 ->internal_coords($ca2, 1.5, $n2, 120, $c, 180);
        $hn2->internal_coords($n2, 1.1, $c, 120, $o, 180)
            if $hn2->symbol eq 'H'; # make sure it's not Proline
        ($c, $o, $ca, $n, $hn) = ($c2, $o2, $ca2, $n2, $hn2);
    }

    sub end_chain {
        my ($chain) = @_;
        my $ox = $chain->new_atom(
            symbol => 'O', 
            internal_coords => [ $c, 1.5, $ca, 120, $o, 180 ]
        );
        my $h = $chain->new_atom(
            symbol => 'H', 
            internal_coords => [ $ox, 1.1, $c, 105, $o, 0 ]
        );
        $chain->new_bond(atoms => [$ox, $c], order => 1);
        $chain->new_bond(atoms => [$ox, $h], order => 1);
        $_->internal_coords->add_cartesians for $chain->atoms;
    }

}

sub safe_clone {
    my ($mol) = @_;
    my $clone = $mol->clone;
    for ($clone, $clone->atoms, $clone->bonds) {
        $_->id($_->nextID);
    }
    $clone;
}

sub build_sequence {
    my ($s) = @_;
    $s or die "no sequence, stopping ";
    my @seq = split //, $s;

    my $code = shift @seq;
    my $res = amino_acid($code);

    my $chain = start_chain($res);
    $chain->name($s);

    for my $code (@seq) {
        $res = amino_acid($code);
        add_residue($chain, $res);
    }
    end_chain($chain);

    return $chain;
}


##############################33

my $chain = build_sequence(shift);
print $chain->print(format => 'mdl');



Makefile.PL
use strict;
use warnings;
#use diagnostics;
use inc::Module::Install;

name('PerlMol');
author('Ivan Tubert-Brohman <itub@cpan.org>');
abstract('Perl modules for chemoinformatics and computational chemistry');
license('perl');
version_from('PerlMol.pm');

requires(qw(
    perl                5.006
    Test::Simple        0.45
    Scalar::Util        1.07
    Storable            2.04
    Text::Balanced      1.89
));

bundle(qw(
    Math::VectorReal                1.02
    Statistics::Regression          0.15
    Parse::Yapp::Driver             1.05
    Chemistry::Mol                  0.32
    Chemistry::MacroMol             0.06
    Chemistry::InternalCoords       0.18
    Chemistry::Ring                 0.18
    Chemistry::Canonicalize         0.10
    Chemistry::Pattern              0.25
    Chemistry::Bond::Find           0.21
    Chemistry::File::MDLMol         0.19
    Chemistry::File::SMILES         0.43
    Chemistry::File::SLN            0.10
    Chemistry::File::PDB            0.20
    Chemistry::FormulaPattern       0.10
    Chemistry::MidasPattern         0.10
    Chemistry::File::Mopac          0.15
    Chemistry::File::SMARTS         0.21
    Chemistry::File::XYZ            0.11
    Chemistry::Isotope              0.10
    Chemistry::Reaction             0.02
    Chemistry::Mok                  0.24
));

check_nmake();

&WriteAll();


